presto-spi/src/main/java/com/facebook/presto/spi/PageBuilder.java

```diff
@@ -43,17 +43,27 @@ public PageBuilder(List<? extends Type> types)
 
     public PageBuilder(int initialExpectedEntries, List<? extends Type> types)
     {
+        this(initialExpectedEntries, DEFAULT_MAX_PAGE_SIZE_IN_BYTES, types);
+    }
+
+    public static PageBuilder withMaxPageSize(int maxPageBytes, List<? extends Type> types)
+    {
+        return new PageBuilder(Integer.MAX_VALUE, maxPageBytes, types);
+    }
+
+    private PageBuilder(int initialExpectedEntries, int maxPageBytes, List<? extends Type> types)
```

Comment:
Not sure I understood your comments. In master we have this code:
```
public PageBuilder(List<? extends Type> types)
    {
        this(Integer.MAX_VALUE, types);
    }
```
so we're already exercising the code path where `initialExpectedEntries` is int_max.

How does setting the block size to `maxPageBytes / types.size()` defeat the purpose of `PageBuilderStatus`? If we set it to something larger the `BlockBuilders` will allocate too much memory. Or did you mean it should be set smaller?

> using max of 32 (when a variable width type exists) and fixed width size as the expected width of the page, instead of sum, doesn't seem to make sense.
I think it does make sense. As soon as any one `BlockBuilder` becomes full the entire `PageBuilder` is full.